### PR TITLE
[MAINT] Store 1 timepoint for 4D generating a report with the maskers

### LINKED ISSUE #3933: Store one time point for reporting that uses 4D images
> the full images are stored for reporting, we should probably only store one time point when the image is 4D

Comes from @jeromedockes comment in https://github.com/nilearn/nilearn/issues/3709#issuecomment-1561608917

### PR DESCRIPTION
**Pull request by ymzayek**
- Closes #3933

Changes proposed in this pull request:

-
-

### CODE REVIEW

**Comment by ymzayek** [comment]
This fails because the multi subject maskers don't support report generation. The same failure can be seen on main if calling `generate_report` on a multi masker object. So the question is do we want to support this for the multi subject maskers? Either way I think the error needs to be improved so I can do that in this PR and can open another issue to see if we would want to support this.

**Comment by bthirion** [comment]
Indeed, we should have report generation for multi subject maskers.

**Comment by bthirion** [comment]
> Ok I tried a few things to make this PR work and not break anything but I think a better strategy might be to better support report generation for the multi subject maskers before doing this. This is because the user could subscript the 5D list and `generate_report` will work as intended and trying to add a warning message about reports in `fit` will probably seem weird to users

Makes sense.

File: nilearn/maskers/multi_nifti_masker.py [hunk 1]
```diff
@@ -240,8 +240,15 @@ def fit(self, imgs=None, y=None):
...
         if self.reports:  # save inputs for reporting
-            imgs = imgs[0] if isinstance(imgs, list) else imgs
-            self._reporting_data = {"images": imgs, "mask": self.mask_img_}
+            self._reporting_data = {"mask": self.mask_img_, "images": imgs}
+            if imgs is not None:
+                imgs = imgs[0] if isinstance(imgs, list) else imgs
+                dim = image.load_img(imgs).shape
+                if len(dim) == 4:
+                    # compute middle image from 4D series for plotting
+                    imgs = image.index_img(imgs, dim[-1] // 2)
```
Comment (bthirion) [inline nilearn/maskers/multi_nifti_masker.py:249]: Can this be factored in a function ? I guess this pattern already exists in the code base ?

File: nilearn/maskers/tests/test_multi_nifti_labels_masker.py [hunk 2]
```diff
@@ -355,6 +355,41 @@ def test_multi_nifti_labels_masker_resampling():
         )
 
 
+def test_multi_nifti_labels_masker_list_of_sample_mask():
```
Comment (bthirion) [inline nilearn/maskers/tests/test_multi_nifti_labels_masker.py]: This is something else than the main topic of this PR (?)

File: nilearn/maskers/tests/test_multi_nifti_maps_masker.py [hunk 3]
```diff
@@ -273,6 +273,39 @@ def test_multi_nifti_maps_masker_resampling():
         assert fmri11_img_r.shape == (masker.maps_img_.shape[:3] + (length,))
 
 
+def test_multi_nifti_maps_masker_list_of_sample_mask():
```
Comment (bthirion) [inline nilearn/maskers/tests/test_multi_nifti_maps_masker.py]: Same question here.

**Review by bthirion — COMMENTED** [review]
Technically, the PR looks good.

File: nilearn/maskers/tests/test_multi_nifti_labels_masker.py (hunk 2 above)
Comment (ymzayek) [reply nilearn/maskers/tests/test_multi_nifti_labels_masker.py]: Hm that's weird. It's not in the diff of the latest commit. Maybe you looked at the diff of the merge commit. I think if you check the latest you won't see it.

File: nilearn/maskers/multi_nifti_masker.py (hunk 1 above)
Comment (ymzayek) [reply nilearn/maskers/multi_nifti_masker.py:249]: Yes probably, let me look into it

**Comment by ymzayek** [comment]
I definitely still need to add tests to this but I think it is finally in good shape for what it's intended to do

**Comment by ymzayek** [comment]
@Remi-Gau have you seen similar failures to the ones on windows here? I can't easily see how they are related but they are consistently failing so there is definitely something there

**Comment by Remi-Gau** [comment]
All of those are related to the use of the `write_tmp_imgs` context manager?

**Comment by Remi-Gau** [comment]
I think that we should probably start removing this context manager in the tests and replace it with a just creating the files in the `tmp_path`.

**Comment by ymzayek** [comment]
Seems to also happen with `write_fake_fmri_data_and_design`. Ok I'll see if rewriting the tests without the context manager resolves this

**Comment by Remi-Gau** [comment]
I am struggling to see what in the PR would start triggering those failures though.

**Comment by ymzayek** [comment]
Ok tests related to `write_tmp_imgs` are not failing anymore with the refactoring. `write_fake_fmri_data_and_design` seems to still be causing issues

**Review by bthirion — APPROVED** [review]
LGTM.

**Review by Remi-Gau — APPROVED** [review]
good with me

**Pull request merged**